consensus/istanbul/core/roundchange.go

```diff
@@ -48,45 +48,125 @@ func (c *core) sendRoundChange(round *big.Int) {
 
 	// Now we have the new round number and sequence number
 	cv = c.currentView()
-	rc := &istanbul.Subject{
-		View:   cv,
-		Digest: common.Hash{},
+	rc := &istanbul.RoundChange{
+		View:                cv,
+		PreparedCertificate: c.current.preparedCertificate,
 	}
 
 	payload, err := Encode(rc)
 	if err != nil {
 		logger.Error("Failed to encode ROUND CHANGE", "rc", rc, "err", err)
 		return
 	}
-
-	c.broadcast(&message{
-		Code: msgRoundChange,
+	logger.Trace("Sending round change message", "rcs", c.roundChangeSet)
+	c.broadcast(&istanbul.Message{
+		Code: istanbul.MsgRoundChange,
 		Msg:  payload,
 	})
 }
 
-func (c *core) handleRoundChange(msg *message, src istanbul.Validator) error {
-	logger := c.logger.New("state", c.state, "from", src.Address().Hex())
+func (c *core) handleRoundChangeCertificate(roundChangeCertificate istanbul.RoundChangeCertificate) error {
+	logger := c.logger.New("state", c.state, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "handleRoundChangeCertificate")
+
+	if len(roundChangeCertificate.RoundChangeMessages) > c.valSet.Size() || len(roundChangeCertificate.RoundChangeMessages) < c.valSet.MinQuorumSize() {
+		return errInvalidRoundChangeCertificateNumMsgs
+	}
+
+	seen := make(map[common.Address]bool)
+	decodedMessages := make([]istanbul.RoundChange, len(roundChangeCertificate.RoundChangeMessages))
+	for i, message := range roundChangeCertificate.RoundChangeMessages {
+		// Verify message signed by a validator
+		data, err := message.PayloadNoSig()
+		if err != nil {
+			return err
+		}
+
+		signer, err := c.validateFn(data, message.Signature)
+		if err != nil {
+			return err
+		}
+
+		if signer != message.Address {
+			return errInvalidRoundChangeCertificateMsgSignature
+		}
+
+		// Check for duplicate ROUND CHANGE messages
+		if seen[signer] {
+			return errInvalidRoundChangeCertificateDuplicate
+		}
+		seen[signer] = true
+
+		// Check that the message is a ROUND CHANGE message
+		if istanbul.MsgRoundChange != message.Code {
+			return errInvalidRoundChangeCertificateMsgCode
+		}
+
+		var roundChange *istanbul.RoundChange
+		if err := message.Decode(&roundChange); err != nil {
+			logger.Error("Failed to decode ROUND CHANGE in certificate", "err", err)
+			return err
+		}
+
+		// Verify ROUND CHANGE message is for a proper view
+		if err := c.checkMessage(istanbul.MsgRoundChange, roundChange.View); err != nil {
+			return errInvalidRoundChangeCertificateMsgView
+		}
+
+		decodedMessages[i] = *roundChange
+	}
+
+	for i, message := range roundChangeCertificate.RoundChangeMessages {
+		// No need to process ROUND CHANGE messages we've seen before.
+		if !c.backend.IsKnownMessage(message) {
+			err := c.handleDecodedCheckedRoundChange(&message, &decodedMessages[i])
+			// We want to continue to process ROUND CHANGE messages if they're for future rounds.
+			// TODO(asa): Should we just process every message regardless of if there's an error?
```

Comment:
Wdyt @trianglesphere?

Reply:
I'm mixed. I think that it could be fine **given** the validator sending the new round is honest, that validator can ensure that it won't fail.

I'm starting to think that we might need some specific logic around `Round Change Certificates` (`view-change` from PBFT) that treats them differently than `round change` or `new view` messages. (In particular pick out the `prepared certificate` with the highest `view` or `round` as the `prepared certificate` that the validator should lock on.